src/utils/__tests__/filterWorkspaces.test.js

```diff
@@ -0,0 +1,62 @@
+import * as path from 'path';
+import { getFixturePath } from 'jest-fixtures';
+
+import filterWorkspaces from '../filterWorkspaces';
+import Project from '../../Project';
+
+/** Checks that a filtered set of workspaces contains all the package names in  expectedWorkspaceNames*/
+function workspacesContains(workspaces, expectedWorkspaceNames) {
+  return expectedWorkspaceNames.every(name => workspaces.some(ws => ws.pkg.config.name === name));
+}
+
+describe('filtering flags', async () => {
+  let allWorkspaces;
+
+  beforeEach(async () => {
+    const fixturePath = await getFixturePath(__dirname, 'simple-project-nested-workspace-no-deps');
+    const project = await Project.init(fixturePath);
+    allWorkspaces = await project.getWorkspaces();
+  });
+
+
+
+  it('should return all workspaces if no filters passed', () => {
+    const filteredWorkspaces = filterWorkspaces(allWorkspaces);
+    expect(filteredWorkspaces).toEqual(allWorkspaces)
+  });
+
+  describe('filtering by name', () => {
+    it('should filter to only names that match --only flag', () => {
+      const filteredWorkspaces = filterWorkspaces(allWorkspaces, { only: 'foo' });
+      expect(filteredWorkspaces.length).toEqual(1);
+      expect(workspacesContains(filteredWorkspaces, ['foo'])).toEqual(true);
+    });
+
+    it('should remove names that match --ignore flag', () => {
+      const filteredWorkspaces = filterWorkspaces(allWorkspaces, { ignore: '@scope/bar' });
+      expect(filteredWorkspaces.length).toEqual(2);
+      expect(workspacesContains(filteredWorkspaces, ['foo', 'baz'])).toEqual(true);
+    });
+
+    it('should support scoped packages (sanity check)', () => {
+      const filteredWorkspaces = filterWorkspaces(allWorkspaces, { only: '**/ba*' });
+      expect(filteredWorkspaces.length).toEqual(2);
+      expect(workspacesContains(filteredWorkspaces, ['@scope/bar', 'baz'])).toEqual(true);
+    })
+
+    it('should using both --only and --ignore', () => {
+      const filteredWorkspaces = filterWorkspaces(allWorkspaces, { only: '**/ba*', ignore: 'baz' });
+      expect(filteredWorkspaces.length).toEqual(1);
+      expect(workspacesContains(filteredWorkspaces, ['@scope/bar'])).toEqual(true);
+    })
+  });
+
+  describe('filtering by path', () => {
```

Comment:
I've run into an issue here.

FilterWorkspaces uses the cwd to find the relative paths to each package so that it can filter by path (i.e --onlyFs="packages/foo"). This is fine when you're actually running commands (assuming you're running at the root of the project), but obviously doesnt work in tests.

There are a couple of solutions... We could pass the cwd to filterWorkspaces... That's a bit messy.

Or maybe we could add relativeDir to each package along with dir? That would be a lot simpler imo and might be useful later on.

Reply:
Going to move into Project 👍